pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,177 @@
+	metric *prometheus.CounterVec
+	*dynamicLabels
+}
+
+func NewCounterVecWithDynamicLabels(counterOpts *prometheus.CounterOpts,
+	labelOptions ...func(labels *dynamicLabels)) *CounterVecWithDynamicLabels {
+	counterVecMetric := &CounterVecWithDynamicLabels{
+		CounterOpts:   counterOpts,
+		dynamicLabels: &dynamicLabels{},
+	}
+	for _, o := range labelOptions {
+		o(counterVecMetric.dynamicLabels)
+	}
+	counterVecMetric.metric = prometheus.NewCounterVec(*counterVecMetric.CounterOpts, counterVecMetric.getLabelList())
+	return counterVecMetric
+}
+
+// RegisterMetricWithDynamicLabels registers the CounterVec metric.
+func (c CounterVecWithDynamicLabels) RegisterMetricWithDynamicLabels() {
+	metrics.Registry.MustRegister(c.metric)
+}
+
+// ObserveMetricWithDynamicLabels logs the counter metric. It simply increments the value, so last parameter isn't used.
+func (c CounterVecWithDynamicLabels) ObserveMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger,
+	fixedLabels map[string]string, _ float64) {
+	labels := defaultMissingLabels(c.getLabelList(), c.buildLabels(fixedLabels, obj))
+	c.metric.With(labels).Inc()
+}
+
+type HistogramVecWithDynamicLabels struct {
+	*prometheus.HistogramOpts
+	metric *prometheus.HistogramVec
+	*dynamicLabels
+}
+
+func NewHistogramVecWithDynamicLabels(histogramOpts *prometheus.HistogramOpts,
+	labelOptions ...func(labels *dynamicLabels)) *HistogramVecWithDynamicLabels {
+	histogramVecMetric := &HistogramVecWithDynamicLabels{
+		HistogramOpts: histogramOpts,
+		dynamicLabels: &dynamicLabels{},
+	}
+	for _, o := range labelOptions {
+		o(histogramVecMetric.dynamicLabels)
+	}
+	histogramVecMetric.metric = prometheus.NewHistogramVec(*histogramVecMetric.HistogramOpts,
+		histogramVecMetric.getLabelList())
+	return histogramVecMetric
+}
+
+// RegisterMetricWithDynamicLabels registers the HistogramVec metric.
+func (h HistogramVecWithDynamicLabels) RegisterMetricWithDynamicLabels() {
+	metrics.Registry.MustRegister(h.metric)
+}
+
+// ObserveMetricWithDynamicLabels observes the histogram metric with val value.
+func (h HistogramVecWithDynamicLabels) ObserveMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger,
+	fixedLabels map[string]string, val float64) {
+	labels := defaultMissingLabels(h.getLabelList(), h.buildLabels(fixedLabels, obj))
+	h.metric.With(labels).Observe(val)
+}
+
+var _ metricsWithDynamicLabels = CounterVecWithDynamicLabels{}
+var _ metricsWithDynamicLabels = HistogramVecWithDynamicLabels{}
+
+// GetOptionalClusterTypeLabels reads the AdditionalClusterDeploymentLabels from the metrics config and returns the same
+// as a map
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) map[string]string {
+	var mapClusterTypeLabelToValue = make(map[string]string)
+	// Currently "cluster_type" is a non-optional label for these metrics, so if the config is empty (default), set the
+	// "cluster_type" label
+	// TODO: remove as a part of https://issues.redhat.com/browse/HIVE-2077
```

Comment:
I think we need to keep the defaulting forever, as we can't guarantee all consumers have cut over to using hiveconfig. Let's just get rid of this comment.